### ISSUE TITLE
pulseaudio daemon not running

### ISSUE THREAD

**Issue by nachochip** [BaReinhard/Multiroom-Audio-iOS#9]
I tested this out with the base Ubuntu for armhf as well as with Ubuntu-Mate for armhf.  Both scenarios came out with a "pulseaudio not running".  Since your software requires the pulseaudio daemon to run under the same user, I added the following line to /etc/rc.local.  This line allows root to run commands as another user.  Once, I rebooted, everything worked fine and gave no errors on the software side.  

runuser -l USERNAME -c 'pulseaudio&'

(I am still waiting on my 2nd chromecast to arrive so I can start making groups, so I can't comment on the sound getting delivered to the chromecasts.  I was under the assumption that single chromecasts could work without being in groups, but the http server page is not displaying any single chromecasts. )

**Comment by BaReinhard** [comment]
It should be visible, it sounds like pulseaudio-dlna may not be running. I would check to see that you have the chromecasts visible as a sink output in pulseaudio

**Comment by nachochip** [comment]
hmmmmm.  And the plot thickens.  So, I came back and noticed that the pulseaudio-dlna was in fact not running.  After implementing the command it argued that "pulseaudio is either not running or dbus module could not be loaded".

So, I think I have deeper issues.  The packages are installed.  I need to tinker some.

**Comment by BaReinhard** [comment]
when did you install this? There was an issue that was causing it with previous installs. I have since updated a few small changes. Additionally, the maintainer of the repo had pushed the latest pulls to the master to the repo so everything should be updated.

What you can try doing is the following command:

```
mv ~/.config/pulse ~/.config/pulse.bad
```
then restart the Pi to see if pulseaudio is now up and running.

What I did find is that sometime when installing from a fresh OS, an occasion pulseaudio has some configuration that doesn't play well, so regenerating it may fix the issue.

**Comment by nachochip** [comment]
I installed within last 48 hours.  Fresh sd card, fresh ubuntu-mate install.  

I will try the command and reply back with results.

**Comment by nachochip** [comment]
ok, I initiated the command (there was already a pulse.bad folder), and rebooted.  I then logged into Ubuntu-Mate and went over to a terminal and noted that it give the same report.  

"pulseaudio_dlna.pulseaudio     CRITICAL  PulseAudio seems not to be running or PulseAudio dbus module could not be loaded.  The application cannot work properly!"

Any suggestions?

**Comment by BaReinhard** [comment]
I would delete the old bad , then re run the command. This is what fixed it for me. I tried several other instances of pulseaudio, that was the only thing that worked. When/if I get some time this weekend I will do a fresh install on a new sd card and try to recreate the issues. This project is really great when it's working I'd hate for you to get turned off because of a small bug I've fixed before.


This time I will be sure to fully document my go round with it.

For the time being can you tell me what os and versions you are using as well as what Pi you are using? 2 or 3

**Comment by nachochip** [comment]
Just deleted pulse.bad folder.  I ran the command again.  Rebooted.  Sad to see the error again.  (Raspberry pi 2.  I historically have used Raspbian for an airprint server for our iOS devices.  I was interested in this Multiroom project and grabbed both the server-only armhf Ubuntu 16.04.2 as well as the Ubuntu-mate armhf 16.04.2)

In the interim, When I login: and re-run the pulseaudio-dlna command it ends up with the same error, but I have found a critical failure right above that last error.   One line above that report is "Failure: Module initialization failed."

I ran `pactl load-module module-dbus-protocol`, and it reported "Failure: Module initialization failed."

I am investigating this for now.

**Comment by nachochip** [comment]
I am trying to troubleshoot.  My investigation let me to 
https://www.raspberrypi.org/forums/viewtopic.php?t=31961&p=278613

This page mentioned to apply the following to /etc/rc.local file

DBUS_SESSION_BUS_ADDRESS=unix:path=/run/dbus/system_bus_socket
DBUS_SESSION_BUS_PID=`cat /run/dbus/pid`
pulseaudio --start

I am still trying to understand where my issue is.  I have only had the Chromecast appear 1 time tonight.  And I rebooted and repeated the steps, trying to get the same results, but not successfull.  I did go through a series of killall pulseaudio and running the pulseaudio-dlna in the userspace instead of in /etc/rc.local (which is where I had seen it inserted).  Will try again and report if I get repeatable results.

**Comment by nachochip** [comment]
ok, one last effort helped me get some repeatable results.  My sequential commands were:

killall pulseaudio
pulseaudio --start
pulseaudio-dlna

Some notes include:
- pulseaudio-dlna had to be run in a terminal in the GUI of Ubuntu-Mate.  It always failed when run in a non-GUI terminal (ctrl+alt+f1, etc.)
- the first two commands could be run in both GUI and non-GUI terminals.
- I ended up removing the pulseaudio-dlna& command from the /etc/rc.local file since it really wasn't helping anything.  Even when rebooting and running the first two commands, it didn't help the currently running pulseaudio-dlna running via /etc/rc.local
- I even left out the DBUS_SESSION.... material mentioned above from my GUI terminal and non-GUI terminal tests.  
- I am still missing any icons on the webpage, but at least I can repeat the Chromecast detection.

Hope this helps adjust any of the scripts that you have in the project.  I will be running a fresh copy of Ubuntu-Mate, pulling the newest git version and seeing if my steps make a difference and report back.

**Comment by BaReinhard** [comment]
This is great, to have it run during boot on a user level you can try adding it to ~/.profile file in your users home directory

**Comment by BaReinhard** [comment]
Additionally, the icons are simply pngs that you will have to download manually for the time being. I believe I will opt for the use of some FontAwesome icon instead of using pngs in my next commit.

**Comment by nachochip** [comment]
So I erased my sd card, started a fresh install of Ubuntu-Mate, grabbed your most recent git repository version (5/20/17). My reported resolution with my 'pulseaudio daemon not running' can be bypassed with the following commands.  I found that running the items you had placed in .profile in it's own script.  I don't plan on logging into the Ubuntu-Mate so here was my experience.

- I installed Ubuntu-Mate
- I installed the git repo and installed multiroom-audio-ios
- I rebooted, and logged in via the GUI of Ubuntu-Mate.  Unfortunately, the GUI login did not complete it's process, just stuck at a menu-less screen with the background showing
- I went to ctl-alt-f1 and noticed the pulseaudio-dlna was not working right and had not detected chromecast-audio.
- I went to ctl-al-f2, logged in, added # to all the associated commands within .profile
- I installed xvfb package
- I created my own file (.multiroom-audio-ios) and added my own virtual x session so it can provide dbus to pulseaudio and work without needing to log into Ubuntu-Mate.
- I added a command to start this new file in rc.local
- I rebooted and without having to log in or intervene, the setup came alive.
- I grabbed my iphone, sent a stream, and 10 seconds later, I heard it on the speakers.
- I never had to use my previous commands mentioned earlier in my posts.

Now my headless usage of Ubuntu-Mate is great.  Two downsides: I lose my tty sessions (ctl-alt-f2, f3, f4, etc) & when I go to log in to Ubuntu-Mate via the GUI login, most of the time it disrupts the output from the Chromecast. This is logical to me now, since the GUI login grabs the dbus and resets the sound output to itself (board audio).   Easy fix though, a person just needs to run changesink (I submitted a pull request to make it nicer when run in a terminal).

I will also place a pull request for the items I mentioned in this post.  Here are the commands though for you to verify.

- Install Ubuntu-Mate, then clone git repo, then install, then reboot, notice it is getting the GUI login stuck, then run these commands.

# get to a new terminal and log in
ctl-alt-f2 
# comment lines out from .profile
nano .profile

	#commented these out
	#/usr/local/bin/autostream&
	#shairport-sync -o alsa -c pulse -c /etc/shairport-sync.conf&
	#nodejs /home/ripena/projects/node/server.js
	#start_web_ui

# install xfvb
sudo apt-get install -y xvfb
# create new script .multiroom-audio-ios  (Of course, putting this in /usr/local/bin would be fine too)
nano ~/.multiroom-audio-ios

	# This command is for running the necessary files for headless usage of multiroom-audio-ios.

	# Start an X server with a virtual framebuffer and let it start
	Xvfb :1 -screen 8 1x1x8 &
	sleep 2

	# Implement dbus on the X instance and let pulseaudio start
	DISPLAY=:1 dbus-launch
	DISPLAY=:1 pulseaudio --start
	sleep 5

	# check the sink inputs and connect them
	/usr/local/bin/autostream&
	sleep 2

	# run airport and web server
	shairport-sync -o alsa -c pulse -c /etc/shairport-sync.conf&
	nodejs /home/ripena/projects/node/server.js
	start_web_ui

# make .multiroom-audio-ios file executable
chmod +x ~/.multiroom-audio-ios

# call this file from rc.local (I'm thinking maybe I should move the .multiroom-audio-ios to /usr/local/bin)
sudo nano /etc/rc.local

	#	I add this after the pulseaudio-dlna&
	runuser -l USERNAME -c '/home/USERNAME/.multiroom-audio-ios'

# Then I rebooted and waited for it to go through it's steps, streamed to the device and heard it 10 seconds later.
sudo reboot

**Comment by BaReinhard** [comment]
Awesome! Thanks for the effort. One thing I might add is that running nodejs for the server.js is the same thing as start_web_ui. So you may want to think about just using one of those.

Also I have just recently pushed a new version of the repo, it fixed some of the hang up on http requests and removed the use of a pngs file and uses font awesome icons in place of the png.

So long as this works I'm all for it. I'd probably prefer to not use an X server so I may have this as another branch until I get some time to test it. I'd prefer to find a fix without an additional dependency, but if that can't be done so be it and I'll push it to the master.

**Comment by nachochip** [comment]
The line start_web_ui existed in the .profile file from the installation process, so I thought it was needed. 
 I will remove start_web_ui and see. Thanks.

I noticed your recent changes, nice.

From my research, it looks like some version of X is needed.  On one hand the xvfb package is ~1,350kB installed size, so it is extremely small.  On the other, I agree with you, without the need for an X server would be great.  I will open a new issue to discuss once I start meddling with headless Ubuntu again to discuss this possibility.  I am going to leave this Ubuntu-mate version running for now, while I meddle with the headless Ubuntu on the side.
